apps/src/lib/tools/jsinterpreter/CustomMarshalingInterpreter.js

```diff
@@ -0,0 +1,272 @@
+const Interpreter = require('@code-dot-org/js-interpreter');
+const codegen = require('../../../codegen');
+
+module.exports = class CustomMarshalingInterpreter extends Interpreter {
+
+  constructor(code, customMarshaler, opt_initFunc) {
+    super(code);
+    if (!customMarshaler) {
+      throw new Error("You must provide a CustomMarshaler to CustomMarshalingInterpreter");
+    }
+    this.customMarshaler = customMarshaler;
+
+    // TODO: because of the way the interpreter is written, we have to re-do
+    // this work from the parent class's constructor because it needs
+    // customMarshaler to have been set before-hand. Unfortunately, javascript
+    // does not allow us to set customMarshaler until after the parent constructor
+    // has finished.
```

Comment:
This TODO does not actually describe an action item.  It should maybe be updated to describe a desired outcome (refactor constructors to do less work, deduplicate code with parent constructor by extracting a helper, etc).